feat: add vlm docs and simple examples

drbh opened the pull request:
This PR start to add documentation for visual language models

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/text-generation-inference/pr_1812). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+
+## What is VLM?
+
+Visual Language Model (VLM) are models that consume both visual and textual inputs to generate text.
```
```suggestion
Visual Language Model (VLM) are models that consume both image and text inputs to generate text.
```
makes it easier to read

drbh replied:
resolved in latests changes

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+
+Visual Language Model (VLM) are models that consume both visual and textual inputs to generate text.
+
+These models are trained on multimodal data, which includes both images and text.
```
would be better not to break the paragraph because you are referring to VLMs with _these_ and the context is the same anyway, this breaks the readability

drbh replied:
paragraph rewritten and is not broken into two sentences in the latest changes

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+
+These models are trained on multimodal data, which includes both images and text.
+
+VLMs can be used for a variety of tasks, such as image captioning, visual question answering, and more.
```
```suggestion
VLMs can be used for a variety of tasks, such as image captioning, visual question answering, and more.
```
these are super common ones that can also be done through models like BLIP that is not often what we mean here with VLMs (models with longer context) so we should add stuff like "multimodal chat" or "visual reasoning" which models like LLaVA is very good at

drbh replied:
mentioned right below the intro

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+    <pre>placeholder for architecture diagram</pre>
+</div>
+
+With VLM, you can generate text from an image. For example, you can generate a caption for an image, answer questions about an image, or generate a description of an image.
```
```suggestion
A vision language model can generate text from an image. For example, you can generate a caption for an image, answer questions about an image, or generate a description of an image.
```
these two sentences are already what you said above

drbh replied:
removed

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+</div>
+
+With VLM, you can generate text from an image. For example, you can generate a caption for an image, answer questions about an image, or generate a description of an image.
+
```
you already mentioned below definitions above, maybe you can remove that and instead say something like:
Below are couple of common use cases for vision language models:

drbh replied:
removed and "Below are couple of common use cases for vision language models:" phrasing used

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+
+- **Image Captioning**: Given an image, generate a caption that describes the image.
+- **Visual Question Answering (VQA)**: Given an image and a question about the image, generate an answer to the question.
+- **Visual Dialog**: Given an image and a dialog history, generate a response to the dialog.
```
```suggestion
- **Multimodal Dialogue**: Generate response to multiple turns of images and conversations.
```
some models can take multiple images at multiple turns without losing context

drbh replied:
`Visual Dialog` removed in favor of this `Multimodal Dialogue` and `Image Information Retrieval`

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+- **Image Captioning**: Given an image, generate a caption that describes the image.
+- **Visual Question Answering (VQA)**: Given an image and a question about the image, generate an answer to the question.
+- **Visual Dialog**: Given an image and a dialog history, generate a response to the dialog.
+- **Visual Data Extraction**: Given an image, extract information from the image.
```
```suggestion
- **Image Information Retrieval**: Given an image, retrieve information from the image.
```
how is this particularly different from visual QA? maybe you might want to specify

drbh replied:
`Visual Dialog` removed in favor of this `Multimodal Dialogue` and `Image Information Retrieval`

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+- **Visual Dialog**: Given an image and a dialog history, generate a response to the dialog.
+- **Visual Data Extraction**: Given an image, extract information from the image.
+
+For example, given the image of a cat, a VLM can generate the caption "A cat sitting on a couch" or answer the question "What is the cat doing?" with "The cat is sitting on a couch."
```
this is VQA IMO, so above case is redundant

drbh replied:
removed

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+
+For example, given the image of a cat, a VLM can generate the caption "A cat sitting on a couch" or answer the question "What is the cat doing?" with "The cat is sitting on a couch."
+
+## How to use VLM?
```
```suggestion
## How to Use a Vision Language Model?
```
consistent casing across headings

drbh replied:
updated

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+
+## How to use VLM?
+
+### VLM with Python
```
```suggestion
### Hugging Face Hub Python Library 
```

drbh replied:
updated

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+
+### VLM with Python
+
+To use VLM with Python, you can use the `huggingface_hub` library. The `InferenceClient` class provides a simple way to interact with the Inference API.
```
```suggestion
To  infer with vision language models through Python, you can use the `huggingface_hub` library. The `InferenceClient` class provides a simple way to interact with the Inference API.
```
maybe give link to documentation of both huggingface_hub, InferenceClient, and Inference API for those who don't know

drbh replied:
link added for `huggingface_hub` and `api-inference` docs

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+
+To use VLM with Python, you can use the `huggingface_hub` library. The `InferenceClient` class provides a simple way to interact with the Inference API.
+
+This is the following image:
```
```suggestion
Let's see an example. We will infer using the image below.
```

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+
+This is the following image:
+
+<div class="flex justify-center">
```
TBH when showing API uses we don't need to see the image as it breaks the flow and is not the main point here

drbh replied:
removed

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
````diff
@@ -0,0 +1,224 @@
+```python
+from huggingface_hub import InferenceClient
+
+client = InferenceClient("http://127.0.0.1:3000")
````
how do people get this up locally though, shouldn't we show this first?

drbh replied:
it appears that the other tutorials/guides do not specify how to start TGI and the examples assume an endpoint is available, and startup instructions are listed in the [quick tour doc](https://huggingface.co/docs/text-generation-inference/quicktour#quick-tour)

However please let me know if I should include startup instructions in this document

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+for token in client.text_generation(prompt, max_new_tokens=16, stream=True):
+    print(token)
+
+# This
```
let's not have this in column-like print

drbh replied:
removed

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+
+Images can be passed as URLs or base64-encoded strings. The `InferenceClient` will automatically detect the image format.
+
+This is the following image:
```
again, no need, we are showing the same API so we can repurpose the same image

drbh replied:
removed

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
````diff
@@ -0,0 +1,224 @@
+# .
+```
+
+Images can be passed as URLs or base64-encoded strings. The `InferenceClient` will automatically detect the image format.
````
we can put this way above IMO as in above example we already passed it as URL

drbh replied:
moved up

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+
+client = InferenceClient("http://127.0.0.1:3000")
+
+url = "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/beaver.png"
```
maybe you could read it from directory this time which is the point of using base64 here

drbh replied:
updated to read image from file

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+
+If you want additional details, you can add `details=True`. In this case, you get a `TextGenerationStreamResponse` which contains additional information such as the probabilities and the tokens. For the final response in the stream, it also returns the full generated text.
+
+### VLM with cURL
```
```suggestion
### Inference Through Sending cURL Requests
```

drbh replied:
updated

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+
+### VLM with cURL
+
+To use the `generate_stream` endpoint with curl, you can add the `-N` flag, which disables curl default buffering and shows data as it arrives from the server
```
Maybe define the endpoint first here as it's the first time you're mentioning it: we don't want to assume the reader knows everything about TGI 

```suggestion
To use the `generate_stream` endpoint with curl, you can add the `-N` flag. This flag disables curl default buffering and shows data as it arrives from the server.
```

drbh replied:
resolved

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
````diff
@@ -0,0 +1,224 @@
+# data:{"index":16,"token":{"id":28723,"text":".","logprob":-0.6196289,"special":false},"generated_text":"This is a picture of an anthropomorphic rabbit in a space suit.","details":null}
+```
+
+### VLM with JavaScript
````
```suggestion
### Inference Through JavaScript
```

drbh replied:
resolved

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+### VLM with JavaScript
+
+First, we need to install the `@huggingface/inference` library.
+`npm install @huggingface/inference`
```
let's have it in three ticks with bash wrapping instead

drbh replied:
resolved

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+First, we need to install the `@huggingface/inference` library.
+`npm install @huggingface/inference`
+
+If you're using the free Inference API, you can use `HfInference`. If you're using inference endpoints, you can use `HfInferenceEndpoint`.
```
What is HfInference? What is HfInferenceEndpoint?

drbh replied:
added short description and link to docs

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
````diff
@@ -0,0 +1,224 @@
+```js
+import { HfInferenceEndpoint } from "@huggingface/inference";
+
+const hf = new HfInferenceEndpoint("http://127.0.0.1:3000", "hf_YOUR_TOKEN");
````
```suggestion
const hf = new HfInferenceEndpoint("http://127.0.0.1:3000", HF_TOKEN);
```

drbh replied:
resolved

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+
+If you're using the free Inference API, you can use `HfInference`. If you're using inference endpoints, you can use `HfInferenceEndpoint`.
+
+We can create a `HfInferenceEndpoint` providing our endpoint URL and credential.
```
```suggestion
We can create a `HfInferenceEndpoint` providing our endpoint URL and [Hugging Face access token](https://huggingface.co/settings/tokens).
```

drbh replied:
resolved

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
````diff
@@ -0,0 +1,224 @@
+// This is a picture of an anthropomorphic rabbit in a space suit.
+```
+
+## Advantages of VLM in TGI
````
```suggestion
## Advantages of Vision Language Models in Text Generation Inference
```
are they specific to TGI in general?

drbh replied:
not exactly any inference server with VLMs and Grammars can support the following example. 

reworded in latest changes

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+
+## How does VLM work under the hood?
+
+coming soon...
```
you don't necessarily need to cover this here, we already have a blog post covering that https://huggingface.co/blog/vlms
we like to keep conceptual guides minimal, although this is not a conceptual guide this is an API guide, it shouldn't be here

drbh replied:
linked

merveenoyan reviewed (commented):
Thanks a lot! 
I think one very major thing is that this isn't a conceptual guide, it should be somewhere in main API reference

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+# Vision Language Models (VLM)
```
```suggestion
# Vision Language Model Inference in TGI
```
In this guide you don't talk about serving but rather inference so better to be specific

drbh replied:
resolved

merveenoyan commented on docs/source/conceptual/visual_language_models.md:
```diff
@@ -0,0 +1,224 @@
+# Vision Language Models (VLM)
+
+## What is VLM?
```
```suggestion
## What is Vision Language Model?
```

drbh replied:
removed

drbh commented:
hi @merveenoyan thank you for the review 🙏 I've moved the doc into the tutorial section and addresses your comments above.

Please let me know if I should make any more changes!
